src/analyses/apron/octApron.apron.ml

```diff
@@ -0,0 +1,192 @@
+      let is = is @ List.map (fun x -> x^"'") is in
+      let fs = fs @ List.map (fun x -> x^"'") fs in
+      let newd = D.add_vars ctx.local (is,fs) in
+      let formargs = Goblintutil.zip f.sformals args in
+      let arith_formals = List.filter (fun (x,_) -> isArithmeticType x.vtype) formargs in
+      List.iter (fun (v, e) -> D.assign_var_with newd (v.vname^"'") e) arith_formals;
+      D.forget_all_with newd (List.map (fun (x,_) -> x.vname) arith_formals);
+      List.iter  (fun (v,_)   -> D.assign_var_eq_with newd v.vname (v.vname^"'")) arith_formals;
+      D.remove_all_but_with newd (is@fs);
+      [ctx.local, newd]
+
+
+  let combine ctx r fe f args fc d =
+    if D.is_bot ctx.local || D.is_bot d then D.bot () else
+      let f = Cilfacade.getdec f in
+      match r with
+      | Some (Var v, NoOffset) when isArithmeticType v.vtype && (not v.vglob) ->
+        let nd = D.forget_all ctx.local [v.vname] in
+        let fis,ffs = D.get_vars ctx.local in
+        let fis = List.map Var.to_string fis in
+        let ffs = List.map Var.to_string ffs in
+        let nd' = D.add_vars d (fis,ffs) in
+        let formargs = Goblintutil.zip f.sformals args in
+        let arith_formals = List.filter (fun (x,_) -> isArithmeticType x.vtype) formargs in
+        List.iter (fun (v, e) -> D.substitute_var_with nd' (v.vname^"'") e) arith_formals;
+        let vars = List.map (fun (x,_) -> x.vname^"'") arith_formals in
+        D.remove_all_with nd' vars;
+        D.forget_all_with nd' [v.vname];
+        D.substitute_var_eq_with nd' "#ret" v.vname;
+        D.remove_all_with nd' ["#ret"];
+        A.unify Man.mgr nd nd'
+      | _ -> D.topE (A.env ctx.local)
+
+  let special ctx r f args =
+    if D.is_bot ctx.local then D.bot () else
+      begin
+        match LibraryFunctions.classify f.vname args with
+        | `Assert expression -> ctx.local
+        | `Unknown "printf" -> ctx.local
+        | `Unknown "__goblint_check" -> ctx.local
+        | `Unknown "__goblint_commit" -> ctx.local
+        | `Unknown "__goblint_assert" -> ctx.local
+        | `Malloc size ->
+          (match r with
+            | Some lv ->
+              D.remove_all ctx.local [f.vname]
+            | _ -> ctx.local)
+        | `Calloc (n, size) ->
+          (match r with
+            | Some lv ->
+              D.remove_all ctx.local [f.vname]
+            | _ -> ctx.local)
+        | `ThreadJoin (id,ret_var) ->
+            let nd = ctx.local in
+            invalidate nd [ret_var];
+            nd
+        | `ThreadCreate _ -> ctx.local
+        | _ ->
+          begin
+            let st =
+              match LibraryFunctions.get_invalidate_action f.vname with
+              | Some fnc -> let () = invalidate ctx.local (fnc `Write  args) in ctx.local
+              | None -> D.topE (A.env ctx.local)
+            in
+              st
+          end
+      end
+
+  let branch ctx e b =
+    if D.is_bot ctx.local then
+      D.bot ()
```

Comment:
why not `raise Deadcode`?

Reply:
To be honest, every one of the transfer functions (already in poly) starts with `if D.is_bot ctx.local then D.bot ()` for some reason. I don't know why this is because transfer functions shouldn't be called on already dead states anyway. For example, we don't do that sort of extra check in base.

Maybe @ivanazuzic knows why (or whether at all) these are necessary.